res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,412 @@
+    midi.sendShortMsg(0x94, 0x1C, 1);                                 // Left Shift Button
+    midi.sendShortMsg(0x95, 0x1C, 1);                                 // Right Shift Button
+};
+
+Prime4.shutdown = function() {
+    // Return all LEDs to initial dim state
+    midi.sendShortMsg(0x90, 0x75, 0x01);
+};
+
+// All components contained in each mixer strip
+const mixerStrip = function(deckNumber, midiChannel) {
+    components.Deck.call(this, deckNumber);
+
+    // PFL Button
+    this.headphoneCue = new components.Button({
+        midi: [0x8F + midiChannel, 0x0D],
+        group: "[Channel" + midiChannel + "]",
+        key: "pfl",
+        on: colDeck[midiChannel - 1],
+        off: colDeckDark[midiChannel - 1],
+        type: components.Button.prototype.types.toggle,
+    });
+
+    this.reconnectComponents(function(c) {
+        if (c.group === undefined) {
+            c.group = this.currentDeck;
+        }
+    });
+};
+
+mixerStrip.prototype = new components.Deck();
+
+// All components contained on each deck
+Prime4.Deck = function(deckNumbers, midiChannel) {
+    components.Deck.call(this, deckNumbers);
+    // Censor Button
+    this.censorButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x01],
+        key: "reverseroll",
+    });
+    // Beatjump Backward
+    this.bjumpBackButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x06],
+        key: "beatjump_backward",
+    });
+    // Beatjump Forward
+    this.bjumpFwdButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x07],
+        key: "beatjump_forward",
+    });
+    // Sync Button
+    this.syncButton = new components.SyncButton({
+        midi: [0x90 + midiChannel, 0x08],
+    });
+    // Cue Button
+    this.cueButton = new components.CueButton({
+        midi: [0x90 + midiChannel, 0x09],
+    });
+    // Play Button
+    this.playButton = new components.PlayButton({
+        midi: [0x90 + midiChannel, 0x0A],
+        unshift: function() {
+            components.PlayButton.prototype.unshift.call(this);
+            this.type = components.Button.prototype.types.toggle;
+        },
+        shift: function() {
+            this.inKey = "play_stutter";
+            this.type = components.Button.prototype.types.push;
+        }
+    });
+    // Hotcue Pads
+    this.hotcuePad = [];
+    for (let i = 1; i <= 8; i++) {
+        this.hotcuePad[i] = new components.HotcueButton({
+            number: i,
+            midi: [0x90 + midiChannel, 0x0E + i],
+            colorMapper: Prime4ColorMapper,
+            off: 0x00,
+            on: 0x03,
+        });
+    }
+    // Tempo Fader
+    this.tempoFader = new components.Pot({
+        midi: [0xB0 + midiChannel, 0x1F],
+        inKey: "rate",
+        invert: true,
+    });
+    // LED indicator when pitch fader is at centre. slightly buggy when switching decks.
+    this.tempoFeedback = function() {
```

Comment:
can you elaborate on the buggyness? This might go away if you model this as a component instead:
```javascript
this.tempoFeedback = new components.Button({
    midi: [0x90 + midiChannel, 0x34],
    outKey: "rate",
    outValueScale: function (value) {
        return value === 0.5 ? this.on : this.off;
    },
})
```
I took the liberty to apply some componentJS "niceness" to get rid of the boilerplate code. Feel free to ask questions if you don't understand my changes. I also simplified the expression because asking whether the rate is `0.5` should work just as well in this particular case (iirc mixxx rounds/snaps the rate to 0.5 and comparing floats to equal 0.5 is also safe because it can be represented exactly unambiguously (unlike some other decimals)).

Reply:
I'm actually not sure what I was talking about with the 'buggyness', as it seems to be a bit smoother now. I would like a way to give the tempo fader a little more bias to being in the middle, as it is currently very difficult to get the original tempo just right, given the faders don't have a bump in the centre, just the light. Thanks for this code though, I'll put it in :) I believe the changes make sense. Seeing you provide these examples certainly helps me understand how Components JS makes things easier, cause it seems to cause me a lot more headaches than anticipated :P

Reply:
> as it is currently very difficult to get the original tempo just right, given the faders don't have a bump in the centre, just the light.

I think you'll just have to live with that. Putting detents into the neutral position of the tempo fader has gone out of fashion, usually because its inaccurate and distracting while also inhibiting fine tempo adjustments around that area. My more expensive Numark NS6II has no detents but a neutral LED (and I think everything above that price category do so too, such as CDJs or your Prime 4). I have a cheaper Reloop controller too that has detents but they're actually quite bad because the hardware is so cheap that they don't line up with the actual centre which is even worse. IIRC there was even some older pro-hardware with a detents/bump in the tempo fader and people modded those to remove them because of the aforementioned issues. 
If you want to reset your tempo fader to neutral, I think you'll just have to dial it in manually by looking at the LED or the BPM display.

Reply:
Yeah fair enough, you're right. Now that the component is mapped properly, it's actually not that bad. Seems just as smooth as using the Prime 4 in standalone mode, which is what I was going for. Thanks for your help :)